index.js

```diff
@@ -224,6 +224,30 @@ export function execaSync(file, args, options) {
 	};
 }
 
+export function $(templatesOrOptions, ...expressions) {
+	if (Array.isArray(templatesOrOptions)) {
+		const [file, ...args] = parseTemplates(templatesOrOptions, expressions);
+		return execa(file, args);
+	}
+
+	return (templates, ...expressions) => {
+		const [file, ...args] = parseTemplates(templates, expressions);
+		return execa(file, args, templatesOrOptions);
+	};
```

Comment:
Should we add a way to bind options multiple times? With each call being shallowly merged to the previous one. This could be useful when someone wants to set some global options once, but then add more options for specific calls. 

Also, when users might be binding options before knowing whether the call will be sync or not. Or they might bind options for multiple calls, sync and async.

For example:

```js
// Global options
$ = $({ env: { ... } })

// ...

// Additional options for a specific call
const result = await $({ shell: true })`ls *`
// And also a sync call
await $.sync({ cwd: '...' })`rm foo`
```

What are your thoughts on this?

Reply:
While this behavior could be well documented, my initial thoughts are that this could also be confusing for users and difficult to provide accurate type definitions.

A similar experience, albeit less elegant, might be encouraging the following workflows:

1. Passing around a global `execaOptions` object
```js
// src/constants.js
export const execaOptions = {...}

// src/script.js
import {execaOptions} from './constants.js'

await $({...execaOptions, shell: true})`ls *`
```

2. Create a project specific wrapper for the `$` API
```js
// src/$.js
import {$ as base$} from 'execa';

const execaOptions = {...}

export const $ = (templatesOrOptions, ...expressions) => {
  if (Array.isArray(templatesOrOptions)) {
    return base$(execaOptions)(templateOrOptions, ...expressions);
  }

  return (templates, ...expressions) => {
    return base$({...execaOptions, ...templatesOrOptions})(templates, ...expressions);
  };
}

// src/script.js
import {$} from './$.js'

await $({shell: true})`ls *`
```

> Note: We could also expose the above recipe as a `create$(baseOptions)` utility

In both cases, I think the code might be easier for users to understand and navigate within their projects. Thoughts?

Reply:
I think the second option might be too much work for users. However, the first option is a possibility.

Yet another possibility would be to use an imperative approach instead of functional, i.e. expose a global mutable `$.options` object that users can modify. I personally dislike that kind of approach because it unnecessarily introduces global state. However, it might be easy to understand and avoid passing option objects around.

I'd like to get @sindresorhus input on this before moving forward. @sindresorhus What do you think?

Summary: how to pass options to `$` tagged template?
  1. `$(options)` binds options. For global options, it can be called several times.
  2. `$(options)` binds options, but can be called only once. For global options, users are expected to pass options objects around.
  3. `$.options` global object.

Reply:
@sindresorhus What do you think?

Reply:
It should definitely be immutable. Options should be passed in normally. Would also be nice to provide an easy API to create a new `$` with custom options (as shown with the "Above example with create$ API" suggestion above). So `2` with a convenience API to create new instances bound to some options.

Reply:
This sounds good, and take the best of both approaches.
What about renaming the following?

```js
// Global options
$ = $({ env: { ... } })
```

To:

```js
// Global options
$ = $.create({ env: { ... } })
```

This removes the need for two different methods `$(options)` and `$.create(options)` which would be quite similar to each other. Like this, users only need to learn a single method `$.create(options)`.

Reply:
Agreed

Reply:
@aaronccasanova What do you think?

Reply:
I like it! I'll do a quick experiment with the `$.create` API and let ya'll know if I hit any blockers (already anticipating some TypeScript complexities).

Some initial thoughts, concerns, clarifications:

**concern:** The examples thus far have been re-assigning the `$` API (e.g. `$ = $.create(...)`). However, from my understanding, this won't really be possible without renaming the imported `$`:
<details>
  <summary>Can't re-assign constant variable <code>$</code></summary>

<img width="904" alt="Screen Shot 2023-02-12 at 10 47 06 AM" src="https://user-images.githubusercontent.com/32409546/218331739-486922f8-8b0c-40ea-b397-2d4d83ced4bc.png">
</details>
<details>
  <summary>Can't import <code>$</code> variable with same name as local identifier</summary>

<img width="899" alt="Screen Shot 2023-02-12 at 11 01 43 AM" src="https://user-images.githubusercontent.com/32409546/218331923-a3cebfce-ac45-4c02-bf0a-1cca1af2d07d.png">
</details>

Perhaps, this is not an issue, but a bit clunky IMO

Note: The `create$` API doesn't have this "issue", but removes the ability to recursively assign and merge options e.g. `$ = $.create(...) = $.create(...) = $.create(...)`

**clarification:** Just want to make sure my previous example is inline with the intention of the API (e.g. `$.create` can be recursively re-assigned and options are shallow merged). I assume so from @sindresorhus's comment: `So each call can extend the existing defaults. ``.create`` can make new defaults be merging with existing ones.`

Reply:
Good catch, you're correct.
So users would need to use a different name and variable than `$` for the return value of `$.create()`. I am actually thinking this might be a good thing? Like this, the original `$` is immutable, and options-bound "copies" of `$` have their own distinct variables, which might create less confusion for users.

```js
import { $ } from 'execa'

const $$ = $.create({ env: { ... } })
await $$({ shell: true })`ls *`
await $$.async({ cwd: '...' })`rm foo`
```

Or:

```js
import { $ as $$ } from 'execa'

const $ = $$.create({ env: { ... } })
await $({ shell: true })`ls *`
await $.async({ cwd: '...' })`rm foo`
```

Reply:
> I am actually thinking this might be a good thing? Like this, the original $ is immutable, and options-bound "copies" of $ have their own distinct variables, which might create less confusion for users.

👍

Reply:
Sounds good 👍 Okay, here's my initial pass at the `$.create` API: https://github.com/sindresorhus/execa/pull/510/commits/5fac1b7ecdafa4b871a81aaeb191cf537116c4ca

I'll pause here for feedback before moving on to types, tests, and docs!

Reply:
While addressing the eslint errors, I noticed and included refactor that removed all the duplicate conditional logic: https://github.com/sindresorhus/execa/pull/510/commits/1bd04ec27d5a055671e7bd8e0983dcde6fd0939e

Reply:
Thanks @aaronccasanova,

The current implementation allows the following two syntaxes:

```js
await $({ shell: true })`ls *`
await $.create({ shell: true })`ls *`
```

I suggested this above, but now I am wondering: those two are actually functionally equivalent, so would it make sense to merge `$(...)` and `$.create(...)` into a single method?

Reply:
I considered this as well, but still liked the idea of a dedicated method for binding options. To play devils advocate, if we merged `$(...)` and `$.create(...)` are there any concerns around creating extraneous functions and memory? Here's a super contrived example:

```js
const $$ = $({ shell: true })

for (const command of commands) {
  await $$(command.options)(...command.args)
}
```

In this example, each `$$(...)` would create a brand new `$` and `$.sync` function. As opposed to calling the same reference returned by `$.create(...)`.

Reply:
When it comes to memory consumption, I think the memory and CPU time wasted by creating anonymous function is quite negligible compared to spawning the child process. For example, CPU time is in the milliseconds-range for child process spawning, while creating anonymous function is in the nanoseconds-range. Memory consumption is also along those lines.

I think having a dedicated method for options binding is nice too, but I am wondering whether there is a technical difference between: 
  - Binding options early (e.g. global options), then spawn multiple processes with it
  - Spawning a single process with specific options

Conceptually, those two could be done with a single method like `$(options)`. Having a single method instead of two could simplify the API for users, and also its implementation. :thinking:

Reply:
Great idea to move `$.sync(...)` to `$(...).sync`! I am wondering how @sindresorhus feels about this, but this look good to me!

Reply:
Okay, here are the two APIs/commits to compare:
- https://github.com/sindresorhus/execa/pull/510/commits/1bd04ec27d5a055671e7bd8e0983dcde6fd0939e: Implements `$.create` for binding options to both the async `$` and `$.sync` APIs and allows one off override options per call e.g. `$(overrideOptions)` and `$.sync(overrideOptions)`
- https://github.com/sindresorhus/execa/pull/510/commits/50055a6a521b5edb8cd805b907e13851a97dbc77: Consolidates option binding and merging to a single method e.g. `$(options)` and `$(options).sync`

Reply:
> Having a single method instead of two could simplify the API for users, and also its documentation, implementation, typing, tests, etc. 

:+1:

> Great idea to move $.sync(...) to $(...).sync! I am wondering how @sindresorhus feels about this, but this look good to me!

:+1:

Reply:
The approach in the current state of the PR (50055a6) looks good to me, thanks again for working on this! :+1: